Importer fails to persist entity due to null non-nullable fields

### Description

Importer failed to persist `entity` objects to db due to null non-nullable fields in integration environment during acceptance tests:

```
2022-08-12T08:54:56.102-0600 ERROR scheduling-6 c.h.m.i.p.r.RecordFileParser Error parsing file 2022-08-12T00_38_26.015902003Z.rcd.gz after 55.53 ms com.hedera.mirror.importer.exception.ParserException: Error copying 14 items to table entity_temp
	at com.hedera.mirror.importer.parser.batch.BatchInserter.persist(BatchInserter.java:112)
	at com.hedera.mirror.importer.parser.batch.CompositeBatchPersister.persist(CompositeBatchPersister.java:62)
	at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.executeBatches(SqlEntityListener.java:299)
	at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.onEnd(SqlEntityListener.java:216)
	at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.onEnd(SqlEntityListener.java:90)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:142)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:52)
	at com.hedera.mirror.importer.parser.AbstractStreamFileParser.parse(AbstractStreamFileParser.java:79)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:117)
	at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:52)
	at com.hedera.mirror.importer.parser.record.RecordFileParser$$FastClassBySpringCGLIB$$8b9310db.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:97)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
	at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:133)
at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:160)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at com.hedera.mirror.importer.parser.record.RecordFileParser$$EnhancerBySpringCGLIB$$6722405b.parse(<generated>)
	at com.hedera.mirror.importer.config.MessagingConfiguration.lambda$integrationFlow$1(MessagingConfiguration.java:107)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.integration.handler.LambdaMessageProcessor.processMessage(LambdaMessageProcessor.java:105)
	at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:105)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:136)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:56)
	at org.springframework.integration.endpoint.PollingConsumer.handleMessage(PollingConsumer.java:158)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.messageReceived(AbstractPollingEndpoint.java:475)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.doPoll(AbstractPollingEndpoint.java:461)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.pollForMessage(AbstractPollingEndpoint.java:413)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.lambda$createPoller$4(AbstractPollingEndpoint.java:348)
	at org.springframework.integration.util.ErrorHandlingTaskExecutor.lambda$execute$0(ErrorHandlingTaskExecutor.java:57)
	at org.springframework.core.task.SyncTaskExecutor.execute(SyncTaskExecutor.java:50)
	at org.springframework.integration.util.ErrorHandlingTaskExecutor.execute(ErrorHandlingTaskExecutor.java:55)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.lambda$createPoller$5(AbstractPollingEndpoint.java:341)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: com.hedera.mirror.importer.exception.ParserException: Error copying 14 items to table entity
	at com.hedera.mirror.importer.parser.batch.BatchUpserter.persistItems(BatchUpserter.java:111)
	at com.hedera.mirror.importer.parser.batch.BatchInserter.persist(BatchInserter.java:108)
	... 54 more
Caused by: org.postgresql.util.PSQLException: ERROR: null value in column "num" of relation "entity" violates not-null constraint
  Detail: Failing row contains (0, 7776000, 1660264707061394499, f, null, 1053, \x1220917cd122311bd200bac97c839c608d284eb32fb01154d119c82dcf3e08..., Mirror Node acceptance test: 2022-08-12T00:38:23.459628124Z Upda..., null, 917cd122311bd200bac97c839c608d284eb32fb01154d119c82dcf3e08d34d6b, null, null, null, null, null, 0, [1660264707061394499,), null, null, 0, \x000000000000000000000000000000000000041d, f, null, -1, -1, null, null, null).
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:181)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:149)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at com.hedera.mirror.importer.parser.batch.BatchUpserter.insertToFinalTable(BatchUpserter.java:117)
	at com.hedera.mirror.importer.parser.batch.BatchUpserter.insertItems(BatchUpserter.java:160)
	at com.hedera.mirror.importer.parser.batch.BatchUpserter.persistItems(BatchUpserter.java:103)
	... 55 more
```

### Steps to reproduce

Run importer built from `main` branch in integration environment and try to parse record file `2022-08-12T00_38_26.015902003Z.rcd.gz`

### Additional context

_No response_

### Hedera network

other

### Version

v0.63.0-SNAPSHOT

### Operating system

_No response_

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListener.java
@@ -608,11 +608,16 @@ public class SqlEntityListener implements EntityListener, RecordStreamFileListen
             dest.setSubmitKey(src.getSubmitKey());
         }
 
-        // There is at least one entity with history. If there is one without history, it must be dest and just copy the
-        // timestamp range from src to dest. Otherwise, both have history, and it's a normal merge from previous to
-        // current, so close the src entity's timestamp range
+        // There is at least one entity with history. If there is one without history, it must be dest and copy non-null
+        // fields and timestamp range from src to dest. Otherwise, both have history, and it's a normal merge from
+        // previous to current, so close the src entity's timestamp range
         if (!dest.isHistory()) {
+            dest.setNum(src.getNum());
+            dest.setRealm(src.getRealm());
+            dest.setShard(src.getShard());
             dest.setTimestampRange(src.getTimestampRange());
+            // It's important to set the type since some non-history updates may have incorrect entity type
+            dest.setType(src.getType());
         } else {
             src.setTimestampUpper(dest.getTimestampLower());
         }

--- a/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListenerTest.java
+++ b/hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/sql/SqlEntityListenerTest.java
@@ -27,6 +27,7 @@ import static com.hedera.mirror.common.domain.entity.EntityType.TOKEN;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.*;
 
+import com.google.common.collect.Range;
 import com.google.protobuf.ByteString;
 import com.hederahashgraph.api.proto.java.Key;
 import java.nio.charset.StandardCharsets;
@@ -377,6 +378,27 @@ class SqlEntityListenerTest extends IntegrationTest {
     }
 
     @Test
+    void onCryptoTransferBeforeContractCreate() {
+        // given
+        var contract = domainBuilder.entity().customize(e -> e.balance(0L).type(CONTRACT)).get();
+        var cryptoTransfer = domainBuilder.cryptoTransfer()
+                .customize(c -> c.amount(10L).consensusTimestamp(contract.getCreatedTimestamp() - 1L)
+                        .entityId(contract.getId()))
+                .get();
+        var expectedContract = TestUtils.clone(contract);
+        expectedContract.setBalance(10L);
+
+        // when
+        sqlEntityListener.onCryptoTransfer(cryptoTransfer);
+        sqlEntityListener.onEntity(contract);
+        completeFileAndCommit();
+
+        // then
+        assertThat(cryptoTransferRepository.findAll()).containsOnly(cryptoTransfer);
+        assertThat(entityRepository.findAll()).containsOnly(expectedContract);
+    }
+
+    @Test
     void onEndNull() {
         sqlEntityListener.onEnd(null);
         assertThat(recordFileRepository.count()).isZero();
@@ -568,6 +590,33 @@ class SqlEntityListenerTest extends IntegrationTest {
 
         assertThat(entityRepository.findAll()).containsExactly(entityMerged);
         assertThat(findHistory(Entity.class)).containsExactly(entity);
+    }
+
+    @Test
+    void onNonHistoryUpdateWithIncorrectTypeThenHistoryUpdate() {
+        // given
+        var entityId = domainBuilder.entityId(ACCOUNT); // The entity in fact is a contract
+        var nonHistoryUpdate = entityId.toEntity();
+        nonHistoryUpdate.setStakePeriodStart(120L);
+        nonHistoryUpdate.setTimestampRange(null);
+        var historyUpdate = entityId.toEntity();
+        historyUpdate.setMemo("Update entity memo");
+        historyUpdate.setTimestampRange(Range.atLeast(domainBuilder.timestamp()));
+        historyUpdate.setType(CONTRACT); // Correct type
+        var expectedEntity = TestUtils.clone(historyUpdate);
+        expectedEntity.setBalance(0L);
+        expectedEntity.setDeclineReward(false);
+        expectedEntity.setEthereumNonce(0L);
+        expectedEntity.setStakedNodeId(-1L);
+        expectedEntity.setStakePeriodStart(120L);
+
+        // when
+        sqlEntityListener.onEntity(nonHistoryUpdate);
+        sqlEntityListener.onEntity(historyUpdate);
+        completeFileAndCommit();
+
+        // then
+        assertThat(entityRepository.findAll()).containsOnly(expectedEntity);
     }
 
     @ParameterizedTest

